FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+# Proposal: Xarray flexible indexes refactoring
+
+Current status: https://github.com/pydata/xarray/projects/1
+
+## 1. Data Model
+
+Indexes are used in Xarray to extract data from Xarray objects using coordinate labels instead of using integer array indices. Although the indexes used in an Xarray object can be accessed (or built on-the-fly) via public methods like `to_index()` or properties like `indexes`, those are mainly used internally.
+
+The goal of this project is to make those indexes 1st-class citizens of Xarray's data model. As such, indexes should clearly be separated from Xarray coordinates with the following relationships:
+
+- Index -> Coordinate: one-to-many
+- Coordinate -> Index: one-to-zero-or-one
+
+An index may be built from one or more coordinates. However, each coordinate must relate to one index at most. Additionally, a coordinate may not be tied to any index.
+
+The order in which multiple coordinates relate to an index should matter. For example, Scikit-Learn's `BallTree` index with the Haversine metric requires providing latitude and longitude values in that specific order. As another example, the order in which levels are defined in a `pandas.MultiIndex` may affect its lexsort depth.
+
+Xarray's current data model is already based on the same index-coordinate relationships. The current data model also implies a one-to-one relationship between a dimension and an index. This one-to-one relationship works as currently it is not possible to perform label-based data selection in Xarray using multi-dimensional coordinates, but since we want to enable this feature in this proposal, we need the dimension-index relationship to evolve towards many-to-many.
```

Comment:
> Xarray's current data model is already based on the same index-coordinate relationships.

I'm not sure I understand what you're trying to say with this statement. Currently there is a one-to-one relationship between and index and a coordinate, not one-to-many. So it seems inconsistent to claim that the " current data model is already based on the same index-coordinate relationships". What do you mean by "the same"? Be more explicit.

Reply:
Right, it is currently one-to-one for index->variable, but it may be one-to-many if we consider that virtual coordinates (multi-index) are coordinates. I can clarify this.